### PR TITLE
Can catch any expression

### PR DESCRIPTION
**Pull request by som-snytt**
`catch` accepts an arbitrary expression, which must conform
to `Function[Throwable, ?]`.

The previous transform was name-based:
```
scala> try 42 catch new {
     | def isDefinedAt(x: Any) = false
     | def apply(x: Any) = ()
     | }
warning: there was one feature warning; re-run with -feature for details
res0: AnyVal = 42

scala> try 42 catch 22
<console>:8: error: value isDefinedAt is not a member of Int
              try 42 catch 22
                           ^
```
Now:
```
scala> try 42 catch 22
                    ^
       error: type mismatch;
        found   : Int(22)
        required: Throwable => ?

scala> def npe: Int = throw null
def npe: Int

scala> val pf: PartialFunction[Throwable, Int] = { case _: NullPointerException => 42 }
val pf: PartialFunction[Throwable,Int] = <function1>

scala> try npe catch pf
val res0: Int = 42

scala> def err: Int = throw new Error()
def err: Int

scala> try err catch pf
java.lang.Error
  at err(<console>:1)
  at liftedTree1$1(<console>:1)
  ... 33 elided
```
As shown, if the handler is a `PartialFunction`, it is invoked
selectively, as though by calling:
```
pf.applyOrElse(e, (t: Throwable) => throw t)
```
Otherwise, the handler is applied to all thrown exceptions:
```
scala> def f(t: Throwable) = if (t.isInstanceOf[NullPointerException]) 27 else ???
def f(t: Throwable): Int

scala> try err catch f
scala.NotImplementedError: an implementation is missing
  at scala.Predef$.$qmark$qmark$qmark(Predef.scala:345)
  at f(<console>:1)
  at $anonfun$res6$1(<console>:1)
  at $anonfun$res6$1$adapted(<console>:1)
  at liftedTree1$1(<console>:1)
  ... 33 elided
```
Note that applying a partial function literal in the form of a pattern-matching
anonymous function will use `applyOrElse`. The following example invokes
the `apply` method because the static type of the handler is not a
`PartialFunction`:
```
scala> val g: Throwable => Int = pf
val g: Throwable => Int = <function1>

scala> try err catch g
scala.MatchError: java.lang.Error (of class java.lang.Error)
  at scala.PartialFunction$$anon$1.apply(PartialFunction.scala:344)
  at scala.PartialFunction$$anon$1.apply(PartialFunction.scala:342)
  at $anonfun$1.applyOrElse(<console>:1)
  at $anonfun$1.applyOrElse(<console>:1)
  at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
  at liftedTree1$1(<console>:1)
  ... 33 elided

```

### CODE REVIEW

**Comment by som-snytt** [comment]
First cut, maybe I cut/pasted wrong. Also, no symbol forcing.

**Comment by som-snytt** [comment]
Rebased. There may be review work pending around hygiene? Actually, it looks like I addressed some comments from a year ago. But there's a bootstrap issue with try relying on library code.

File: src/compiler/scala/tools/nsc/ast/parser/TreeBuilder.scala
Comment (retronym) [inline src/compiler/scala/tools/nsc/ast/parser/TreeBuilder.scala]: This appears unhygenic, 😷

File: src/compiler/scala/tools/nsc/ast/parser/TreeBuilder.scala
Comment (retronym) [inline src/compiler/scala/tools/nsc/ast/parser/TreeBuilder.scala]: The parser needs to be threadsafe, so this would need to be eagerly initialized. We eagerly initialize all class/module symbols referenced in the parser in `forceSymbolsUsedByParser`. See 760df9843a910d6c3618e490c752eb03fb6924bd for the corresponding test to drive the change.

I would probably suggest selecting the method by name (`Select(ScalaRuntimeModule, nme.catchIdentity))` here to avoid finding the right spot to initialize the lazy val in `RunDefinition`s in the same manner as  `forceSymbolsUsedByParser`. It is probably safest not to use `Run` at all from the parser.

File: src/compiler/scala/tools/nsc/ast/parser/TreeBuilder.scala
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/ast/parser/TreeBuilder.scala]: I think you mentioned that a year ago... If I can get it to build, I'll polish it.

File: src/compiler/scala/tools/nsc/ast/parser/TreeBuilder.scala
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/ast/parser/TreeBuilder.scala]: Thanks, I remember I hadn't understood this yet.

**Comment by retronym** [comment]
Would it be possible to submit the part of this that fixes parsing of `try { 1; 1 } + 1 catch` from the part that changes how catch expressions are translated?  I can't remember the context for the latter part, and I'm a bit hesitant about having `catch` sometimes translating into something heavyweight (an `applyOrElse` call with a lambda as an argument).

**Comment by som-snytt** [comment]
I'll refresh my memory when I'm awake again.

**Comment by SethTisue** [comment]
7-year itch, scratched.

**Comment by som-snytt** [comment]
7 years later.

**Comment by dwijnand** [comment]
:head_scratch:  is this addressing https://github.com/scala/bug/issues/5887?

**Comment by som-snytt** [comment]
@dwijnand this is the catch part. It doesn't typecheck correctly yet, mea culpa. Dotty requires a Function instead of PartialFunction, but this is an improvement over the status quo.

**Comment by dwijnand** [comment]
Closing for inactivity.

**Comment by som-snytt** [comment]
I thought drafts were immune from harassment. Since they got rid of "WIP" label, there's no way to rediscover a closed PR in the pile.

**Comment by SethTisue** [comment]
@som-snytt how would you feel about using a "revisit" label for this? that's what they have over in the Dotty repo: https://github.com/lampepfl/dotty/issues?q=label%3Astat%3Arevisit+is%3Aclosed

**Comment by SethTisue** [comment]
for context, for the triage crew of me and Dale, the difference between "open draft" and "closed with revisit label" would be that we make the rounds of all open PRs every month or so, or even more often than that, and see if they need attention. closing as "revisit"  would mean the triage crew is unlikely to look at it again unless a contributor says/does something

**Comment by som-snytt** [comment]
I can self-assign as a marker on my page of pulls. As opposed to a label.

(It's hard enough to remember my two or three current tasks for my actual job, let alone casual contributions from months ago.)

**Comment by SethTisue** [comment]
okay, let's try that for now (and I hope I've understood correctly that that means we can close a PR like this that hasn't been active for a couple of months)

...but if you end up concluding that adding and using a "revisit" label would be better overall, don't hesitate to speak up

**Comment by som-snytt** [comment]
I'm about to push to this branch...

**Comment by SethTisue** [comment]
careful, you'll give ammunition to sneaky maintainers whose real motive in closing a PR is to provoke the contributor into reviving their changes 😈

**Comment by som-snytt** [comment]
I should have commented, Opening for activity.

Thanks for everyone's patience.

**Comment by lrytz** [comment]
WWDD? I ran a few examples, and it seems the same as this PR, so this is awesome. Oh and you already point that out in the PR description.

LGTM!

@retronym maybe you'd like to take another look, since you've been here 4½ years ago?

The breaking change is that the name-based translation no longer kicks in. In `try e catch h`, if `h` is statically conforms to `PartialFunction`, then `isDefinedAt` is checked. Otherwise, `h` needs to conform to `Throwable => ?`.

```scala
scala> object h1 { def isDefinedAt(x: Any) = true; def apply(x: Any) = 1 }
scala> val h2: PartialFunction[Throwable, Int] = { case _: NullPointerException => 1 }
scala> val h3: Throwable => Int = h2

scala> def e = throw new Error()

// used to be allowed before
scala> try e catch h1
                   ^
       error: type mismatch;
        found   : h1.type
        required: Throwable => ?

// same as before
scala> try e catch h2
java.lang.Error
  at e(<console>:1)
  at liftedTree1$1(<console>:1)
  ... 32 elided

// used to not compile
scala> try e catch h3
scala.MatchError: java.lang.Error (of class java.lang.Error)
  at scala.PartialFunction$$anon$1.apply(PartialFunction.scala:344)
  at scala.PartialFunction$$anon$1.apply(PartialFunction.scala:342)
  ... 32 elided
```

Dotty does the same for these examples, so 👍. Should these be added as test cases?

**Comment by som-snytt** [comment]
Added the clever tests.

I forgot to use the swanky syntax from the old PR:
```
scala> try 42 catch new {
     | def isDefinedAt(x: Any) = false
     | def apply(x: Any) = ()
     | }
```

If schools were open, this PR would be entering Kindergarten next fall.

**Comment by som-snytt** [comment]
The previous idea was to desugar `try e catch h` as
```
try e catch {
  case t => pf(h).applyOrElse(t, e => throw e)
}
```
as opposed to
```
try e catch {
  case t =>
    val catchExpr$1 = h
    if (catchExpr$1.isDefinedAt(t)) catchExpr$1.apply(t)
    else throw t
}
```
I'm not sure why `applyOrElse` did not survive in the current iteration. It probably doesn't matter much, since catches don't usually have many cases, and usually simple type tests.

**Comment by SethTisue** [comment]
@som-snytt it appears this will be merged, so you can bring the PR description over from the old PR (and update it if needed, make sure the text is suitably user-facing and not just reviewer-facing...)

File: test/files/neg/t5887.scala
Comment (dwijnand) [inline test/files/neg/t5887.scala:12]: This one caught my eye.  Using `case` we get a warning:

```scala
scala> try ??? catch { case _ => 42 }
                            ^
       warning: This catches all Throwables. If this is really intended, use `case _ : Throwable` to clear this warning.
val res2: Int = 42
```

Could get the same warning for your example?  Not a merge blocker IMO.

File: test/files/neg/t5887.scala
Comment (dwijnand) [inline test/files/neg/t5887.scala]: stale

**Comment by som-snytt** [comment]
Some people set a high bar, others set a high loop to jump through.

Squashed and rebased final edit.

@SethTisue I couldn't bring myself to delete my comment so I moved it here:

The catcher in the wry. Bring parody with dotty.

Used the Lukas examples in the description as most descriptive.

Update: also tweaked the spec and squashed again.

File: test/files/neg/t5887.scala
Comment (som-snytt) [reply test/files/neg/t5887.scala:12]: "It's just a function."

How about:
```
warning: This is a function. If this is really intended, it should be enabled by importing language.functionalProgramming
```

File: test/files/neg/t5887.scala
Comment (som-snytt) [reply test/files/neg/t5887.scala]: Dale! oh right, as Lukas hinted, it expects a function but will leverage a PF. I'll update the description accordingly.

File: test/files/neg/t5887.scala
Comment (dwijnand) [reply test/files/neg/t5887.scala:12]: Catching OutOfMemoryExceptions (and then allocating) is lint worthy, IMO...

Comment (som-snytt) [reply test/files/neg/t5887.scala:12]: Apparently my previous joke is not a joke. I think you'd have to lint the feature. But if a catch function is just a dangerous niche feature, then it seems to me that it doesn't pull its weight. Maybe it should lint `try`.
```
Did you mean `Try`?
```

**Comment by som-snytt** [comment]
Incredibly, after updating ten days ago, there's a spec conflict.

**Comment by SethTisue** [comment]
Assuming one agrees the change is desirable, then this looks ready to merge.

The lack of a lint bothers me a little, since if it was worth linting in the `PartialFunction` case, why isn't it worth also listing int his case?

But actually my doubt is even broader than that: don't understand why this change, in its entirety, would be considered desirable. Allowing `Function` instead of requiring `PartialFunction` only encourages catching `Throwable`, which is pretty much never a good idea (because of e.g. `OutOfMemoryError`). What _desirable_ code pattern does this enable?

I see that Dotty already does the same thing, but I would like to know it was even _intentional_ that Dotty accepts it, and if so, what the rationale was.

**Comment by dwijnand** [comment]
Is https://github.com/lampepfl/dotty/pull/8664 where this was opened up in Scala 3?  @odersky do you think it's risky to allow catching any Throwable in Scala 3 in a `catch`?

**Comment by som-snytt** [comment]
My previous joke about `Try` is relevant. That is, prefer `Try` to `try` for this sort of safety.

I'll add a lint if that will assuage our collective conscience, but I think the catch function is quite niche.

The use case for linting would be, when I write the handler, they never made it easy for me to specify the type:

```
val handler: PartialFunction[Throwable, String] = { case _ => "hello, world" }

val danger_? = (t: Throwable) => "goodbye, cruel world"

try e catch ???
```
Using `handler` doesn't warn. Or rather, defining it does not warn.

**Comment by som-snytt** [comment]
Gah, lost my comment edit when I clicked on the validate failure! which was spurious, of course.

To recap, now it issues the usual warning on catch with total function.

My comment also complained about stretching this PR into a long tail.

To answer Seth's question, the usual tractable syntax is the block of case clauses. Anyone doing something else syntactically will know what they're doing.

I wanted to make the case that in a Scalable language, I should be able to say `try f() catch println` in REPL. But I don't want to get into a philosophical debate about it either.

This change goes to regularity.

**Comment by SethTisue** [comment]
@lrytz care to weigh in about the design issues here?

**Comment by som-snytt** [comment]
Closing because I don't care any more.

**Comment by lrytz** [comment]
> Allowing `Function` instead of requiring `PartialFunction` only encourages catching `Throwable`, which is pretty much never a good idea (because of e.g. `OutOfMemoryError`). What _desirable_ code pattern does this enable?

This is a good point.

Opened https://github.com/lampepfl/dotty/issues/11019, also that Scala 3 doesn't have the "catches all Throwables" warning at all.

**Comment by som-snytt** [comment]
The good point was addressed in the follow-up.

**Comment by lrytz** [comment]
👍, I noticed the warning. So it would be a new, handy feature for yolo-hacking, but discouraged from the beginning for larger projects.

**Comment by SethTisue** [comment]
(To be clear, I'm not _opposed_, as long as it warns. But can't get excited about it either.)

**Review by dwijnand — APPROVED** [review]
Thanks, Som!

**Comment by lrytz** [comment]
I'm slightly torn; it's generally good to align Scala 2 and 3, and I also think having the warning is right. But then we introduce a new feature that gives a warning as soon and as long as you use it. Does that make sense?

Are we certain that the change in dotty was fully intentioned?

**Comment by som-snytt** [comment]
We can adjust later to the status quo established by the dotty ticket.

It's not like everyone will start writing `try e catch f` everywhere. Probably no one notices the difference.

But it's much more regular than the previous state of affairs.

I had an example of a feature that was introduced and immediately deprecated, but I don't remember it off the cuff.

For the record, now I believe the compiler should emit no warnings at all, because someone will always complain either that the warning is too strict or too lax. Leave it to 3rd party tooling to express preferences and unwarranted fears about style.

I re-opened this PR because it's an obvious good.

**Comment by SethTisue** [comment]
there hasn't been any response on https://github.com/lampepfl/dotty/issues/11019

Lukas and I are both on the fence, but Dale likes it, and Som is still behind this even after having had six years to think about it (😬), so let's merge

**Comment by som-snytt** [comment]
I'm going to delete this branch and go back to sleep.

I'm grateful there was no need to come up with an option name like for splain.

I'm sad Seth is not excited.

I see there were a few previous summations. My footnote would be to echo my previous comment: really, people on the fence want to say, Don't use try, use Try instead. Why don't we say that?

Because of the pun, "on the fens", I have learned from wikipedia that a fen is a type of mire, but less acidic than a bog. Please keep this in mind when saying casually, "I'm bogged down," "in a mire," etc, let alone, "on the fens." Let's use language precisely.

**Comment by SethTisue** [comment]
> a fen is a type of mire, but less acidic than a bog

as for me, just yesterday I learned what “muskeg” is

**Comment by som-snytt** [comment]
This PR was definitely muskegged.

No relation to "musk egg," "Muskogee," or "beer keg." The keg hauled by moose on rescue missions would be moose keg.

**Pull request merged**
